File: pi/logic/terms.py

#
# Copyright 2009 Eigenlabs Ltd.  http://www.eigenlabs.com
#
# This file is part of EigenD.
#
# EigenD is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# EigenD is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with EigenD.  If not, see <http://www.gnu.org/licenses/>.
#

import string

__tx = string.maketrans('','')

def make_subst(word):
    return Subst(word)

def render_termlist(tl):
    return ','.join(map(render_term,tl))

def render_result(e):
    return ','.join([ "%s=%s" % (v, render_term(e[v])) for v in e ])

def render_term(t):
    if t is None: return 'None'
    if is_term(t): return str(t)
    if is_variable(t): return str(t)
    if is_split(t): return str(t)
    if is_list(t): return '[%s]' % render_termlist(t)
    if isinstance(t,Subst): return '~%s' % t.word
    if isinstance(t,str): return quotesimplename(t)
    if isinstance(t,int): return str(t)
    if isinstance(t,bool): return str(t)
    if isinstance(t,long): return str(t)
    if isinstance(t,float): return str(t)
    if isinstance(t,Expansion): return t.pred
    return '<%s>' % repr(t)

def make_rule(h,*g):
    return Rule(h,g)

def make_term(t,*a):
    return Term(str(t),*a)

def make_expansion(p):
    return Expansion(p)

def make_variable(v):
    return Variable(str(v))

def make_split(h,t):
    if is_bound(h) and is_bound(t):
        return (h,)+t
    return Split(h,t)

def is_pred(term,pred):
    return is_term(term) and term.pred==pred

def is_pred_arity(term,pred,min,max=None):
    if is_pred(term,pred):
        if term.arity >= min:
            if max is None or term.arity<=max:
                return True
    return False

def is_term(term):
    return isinstance(term,Term)

def is_split(term):
    return isinstance(term,Split)

def is_bound(term):
    if isinstance(term,Variable): return False
    if isinstance(term,Term): return term.bound
    if isinstance(term,Split): return False
    if is_list(term):
        for i in term:
            if not is_bound(i):
                    return False
    return True

def is_unbound(term):
    return not is_bound(term)

def is_variable(term):
    return isinstance(term,Variable)

def is_atom(term):
    if isinstance(term,Variable): return False
    if isinstance(term,Term): return False
    if isinstance(term,Split): return False
    if is_list(term): return False
    return True

def is_const(term):
    if isinstance(term,Variable): return False
    if isinstance(term,Term): return False
    if isinstance(term,Split): return False
    if is_list(term): return False
    return True

def is_list(term):
    return isinstance(term,tuple)

def expand(term,env):
    if is_bound(term):
        return term

    if is_variable(term):
        if term.pred in env:
            return expand(env[term.pred],env)
        return term

    if is_split(term):
        return make_split(expand(term.head,env),expand(term.tail,env))

    if is_list(term):
        return tuple( expand(arg,env) for arg in term )

    if term.arity==0:
        return term

    args = [ expand(arg,env) for arg in term.args ]
    return make_term(term.pred, *args)

def match(left,right,env):
    env.clear()
    return unify(left,{},right,env)

def unify(src, src_env, dest, dest_env):
    while True:
        if is_variable(src):
            if not src.pred in src_env:
                return True
            src = src_env[src.pred]
            continue

        if is_variable(dest):
            if not dest.pred in dest_env:
                dest_env[dest.pred] = expand(src,src_env)
                return True
            dest = dest_env[dest.pred]
            continue

        if is_bound(src) and is_bound(dest):
            return src==dest

        if is_const(src):
            if not is_const(dest):
                return False
            return src==dest

        if is_const(dest):
            if not is_const(src):
                return False
            return src==dest

        if is_split(src):
            if is_split(dest):
                dest_copy = dest_env.copy()
                if not unify(src.head,src_env,dest.head,dest_copy): return False
                if not unify(src.tail,src_env,dest.tail,dest_copy): return False
                dest_env.update(dest_copy)
                return True
            if is_list(dest) and len(dest)>0:
                dest_copy = dest_env.copy()
                if not unify(src.head,src_env,dest[0],dest_copy): return False
                if not unify(src.tail,src_env,dest[1:],dest_copy): return False
                dest_env.update(dest_copy)
                return True
            return False

        if is_split(dest):
            if is_list(src) and len(src)>0:
                dest_copy = dest_env.copy()
                if not unify(src[0],src_env,dest.head,dest_copy): return False
                if not unify(src[1:],src_env,dest.tail,dest_copy): return False
                dest_env.update(dest_copy)
                return True
            return False

        if is_list(src):
            if not is_list(dest): return False
            if len(src) != len(dest): return False
            dest_copy = dest_env.copy()
            for (s,d) in zip(src,dest):
                if not unify(s,src_env,d,dest_copy): return False
            dest_env.update(dest_copy)
            return True

        if is_list(dest):
            return False

        if src.signature() != dest.signature():
            return False

        dest_copy = dest_env.copy()
        for i in range(src.arity) :
            if not unify(src.args[i],src_env,dest.args[i],dest_copy) : return False
        dest_env.update(dest_copy)
        return True

def quotevarname(name):
    h=name[0].translate(__tx,'ABCDEFGHIJKLMNOPQRSTUVWXYZ_')
    t=name[1:].translate(__tx,'ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz0123456789')
    if not h and not t: return name
    return "%%'%s'" % name.replace('%','%25').replace("'","%27")

def quotesimplename(name):
    if name.startswith('~'): return name
    if name is True: return 'True'
    if name is False: return 'False'
    if isinstance(name,int): return str(name)
    if isinstance(name,long): return str(name)
    if isinstance(name,float): return str(name)
    if not isinstance(name,str): return '<%s>' % str(name)
    if not name: return "''"
    h=name[0].translate(__tx,'abcdefghijklmnopqrstuvwxyz$@!#')
    t=name[1:].translate(__tx,'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789#_.')
    if not h and not t: return name
    return "'%s'" % name.replace('%','%25').replace("'","%27")

class Expansion(object):
    __slots__ = ('pred')

    def __init__(self, pred):
        self.pred = pred

    def __str__(self):
        return self.pred

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return hash(self.pred)

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Expansion): return 1
        return cmp(self.pred,other.pred)


class Variable(object):
    __slots__ = ('pred')

    def __init__(self, pred):
        self.pred = pred

    def __str__(self):
        return quotevarname(self.pred)

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return hash(self.pred)

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Variable): return 1
        return cmp(self.pred,other.pred)

class Split(object):
    __slots__ = ('head','tail','hash')

    def __init__(self,head,tail):
        self.head = head
        self.tail = tail
        self.hash = hash(head)^hash(tail)

    def __str__(self):
        return '[%s|%s]' % (self.head,self.tail)

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return self.hash

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Split): return 1
        x = cmp(self.head,other.head)
        if x != 0: return x
        return cmp(self.tail,other.tail)

class Term(object):
    __slots__ = ('pred','args','bound','hash','arity','isig')

    def __init__(self, pred, *args):
        self.pred = pred
        self.args = args
        self.bound = True
        self.hash=13

        self.isig="%s/%u" % (pred,len(args))
        self.arity=len(args)

        if pred.startswith('$'):
            self.bound = False

        for a in args:
            self.hash=self.hash^hash(a)
            if is_unbound(a):
                self.bound=False

        self.hash=self.hash^hash(pred)
        self.hash=self.hash^hash(self.bound)

    def __ne__(self,other):
        return not self.__eq__(other)

    def __eq__(self,other):
        if self is other: return True
        if not isinstance(other,Term): return False
        if self.arity != other.arity: return False
        if self.pred != other.pred: return False
        if not self.args: return True
        for (s,o) in zip(self.args,other.args):
            x = (s==o)
            if not x:
                return False
        return True

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Term): return 1
        if self.arity < other.arity: return -1
        if self.arity > other.arity: return 1
        c = cmp(self.pred,other.pred)
        if c!=0: return c
        if not self.args: return 0
        for (s,o) in zip(self.args,other.args):
            x = cmp(s,o)
            if x != 0: return x
        return 0

    def __str__(self):
        p = self.pred
        a = self.arity
        return "%s(%s)" % (p,','.join(map(render_term,self.args)))

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return self.hash

    def signature(self):
        return self.isig

class Rule:

    __slots__ = ('head','goals','hash','count')

    def __init__(self, head, goals=()):
        self.head = head
        self.goals = goals
        self.count = 0

        h=hash(head)
        for a in goals: h=h^hash(a)
        self.hash=h

    def __hash__(self):
        return self.hash

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Rule): return -1
        if self.head != other.head: return -1
        if len(self.goals) != len(other.goals): return -1
        for (s,o) in zip(self.goals,other.goals):
            x=cmp(s,o)
            if x != 0:
                return x
        return 0

    def __str__(self):
        if len(self.goals)==0: return "%s." % self.head
        return "%s:-%s." % (self.head,','.join(map(str,self.goals)))

    def __repr__(self):
        return self.__str__()

class Subst:
    def __init__(self,word):
        self.word=word

    def __str__(self):
        return self.word

    def __repr__(self):
        return self.__str__()

    def __hash__(self):
        return hash(self.word)

    def __cmp__(self,other):
        if self is other: return 0
        if not isinstance(other,Subst): return 1
        return cmp(self.word,other.word)

File: app_stage/Terms.cpp
/*
 Copyright 2010-2014 Eigenlabs Ltd.  http://www.eigenlabs.com

 This file is part of EigenD.

 EigenD is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.

 EigenD is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with EigenD.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "Terms.h"
#include <ctype.h>
#include <map>

struct term_t::impl_t: public ReferenceCountedObject
{
    impl_t(unsigned type): type_(type) { }
    virtual ~impl_t() { }

    virtual String pred() const { return String::empty; }
    virtual String value() const { return String::empty; }
    virtual unsigned arity() const { return 0; }
    virtual term_t arg(unsigned) const { return term_t(); }
    virtual void set_arg(unsigned,const term_t &) {}
    virtual void append_arg(const term_t &v) {}
    virtual void render(String &) const = 0;

    unsigned type_;
};

namespace
{
    struct pd_t
    {
        virtual ~pd_t() {}
        virtual void destroy_any(void *) = 0;
        virtual void *make_collection() = 0;
        virtual void add_arg(void *, void *) = 0;
        virtual void *make_term(const char *, unsigned, void *) = 0;
        virtual void *make_list(void *) = 0;
        virtual void *make_split(void *, void *) = 0;
        virtual void *make_long(long) = 0;
        virtual void *make_float(float) = 0;
        virtual void *make_bool(bool) = 0;
        virtual void *make_none() = 0;
        virtual void *make_string(const char *, unsigned) = 0;
        virtual void *make_variable(const char *, unsigned) = 0;
        virtual void *make_subst(const char *, unsigned, const char *, unsigned) = 0;
    };

    void render_string(String &s, const char *str)
    {
        s << '\'';

        while(*str)
        {
            switch(*str)
            {
                case '\'': s << "%27"; break;
                case '%': s << "%25"; break;
                default: s << *str; break;
            }

            str++;
        }

        s << '\'';
    }

    char to_hex(unsigned d)
    {
        if(d<10) return '0'+d;
        return d-10+'a';
    }

    struct term_atom_t: term_t::impl_t
    {
        term_atom_t(const String &value): impl_t(PIW_TERM_ATOM), value_(value) {}
        ~term_atom_t() {}
        String value() const { return value_; }

        void render(String &stream) const { stream << value_; }

        String value_;
    };

    struct term_list_t: term_t::impl_t
    {
        term_list_t(unsigned len): impl_t(PIW_TERM_LIST), values_(len) {}
        term_list_t(): impl_t(PIW_TERM_LIST) {}
        ~term_list_t() { values_.resize(0); }

        unsigned arity() const { return values_.size(); }
        term_t arg(unsigned n) const { return values_[n]; }
        void set_arg(unsigned n,const term_t &v) { values_[n] = v; }

        void append_arg(const term_t &v)
        {
            unsigned n = values_.size();
            values_.resize(n+1);
            values_[n].reset();
            values_[n]=v;
        }

        void render(String &stream) const
        {
            stream << '[';
            bool comma = false;

            for(unsigned i=0;i<values_.size();i++)
            {
                const impl_t *impl = values_[i].impl();

                if(comma) stream << ',';
                comma = true;

                if(impl)
                {
                    impl->render(stream);
                }
                else
                {
                    stream << "None";
                }
            }

            stream << ']';
        }

        std::vector<term_t> values_;
    };

    struct term_pred_t: term_t::impl_t
    {
        term_pred_t(const String &pred, unsigned len): impl_t(PIW_TERM_PRED), pred_(pred), values_(len) { }
        term_pred_t(const String &pred): impl_t(PIW_TERM_PRED), pred_(pred) { }
        ~term_pred_t() { values_.resize(0); }

        unsigned arity() const { return values_.size(); }
        term_t arg(unsigned n) const { return values_[n]; }
        void set_arg(unsigned n,const term_t &v) { values_[n] = v; }

        void append_arg(const term_t &v)
        {
            unsigned n = values_.size();
            values_.resize(n+1);
            values_[n].reset();
            values_[n]=v;
        }

        String pred() const { return pred_; }

        void render(String &stream) const
        {
            stream << pred_ << '(';
            bool comma = false;

            for(unsigned i=0;i<values_.size();i++)
            {
                const impl_t *impl = values_[i].impl();

                if(comma) stream << ',';
                comma = true;

                if(impl)
                {
                    impl->render(stream);
                }
                else
                {
                    stream << "None";
                }
            }

            stream << ')';
        }

        String pred_;
        std::vector<term_t> values_;
    };

    void *parse_clause_(pd_t *pd, const char **in);

    void skip_space(const char **in)
    {
        while(**in && isspace(**in))
        {
            (*in)++;
        }
    }

    void skip_one(const char **in)
    {
        if(**in) 
        {
            (*in)++;
        }
    }

    bool is_body(char w)
    {
        if(!w || w==')' || w=='|' || w==',' || w==']') return false;
        return true;
    }

    bool is_string_start(char w)
    {
        if(w=='\"' || w=='\'') return true;
        return false;
    }

    bool is_number_start(char w)
    {
        if(w>='0' && w<='9') return true;
        if(w=='.') return true;
        if(w=='-') return true;
        if(w=='+') return true;
        return false;
    }

    bool is_variable_start(char w)
    {
        if(w>='A' && w<='Z') return true;
        if(w=='_') return true;
        return false;
    }

    bool is_variable_body(char w)
    {
        if(w>='a' && w<='z') return true;
        if(w>='A' && w<='Z') return true;
        if(w>='0' && w<='9') return true;
        if(w=='_') return true;
        return false;
    }

    bool is_word_start(char w)
    {
        if(w>='a' && w<='z') return true;
        if(w>='0' && w<='9') return true;
        if(w=='!' || w=='$' || w=='@' || w=='#' || w=='<' || w=='>') return true;
        return false;
    }

    bool is_word_body(char w)
    {
        if(is_word_start(w)) return true;
        if(w>='A' && w<='Z') return true;
        if(w=='_' || w=='.') return true;
        return false;
    }

    void *parse_list_(pd_t *pd, const char **in)
    {
        skip_space(in);

        if(**in != '[')
        {
            return 0;
        }

        skip_one(in);

        skip_space(in);

        if(**in == ']')
        {
            skip_one(in);
            return pd->make_list(pd->make_collection());
        }

        void *t1 = parse_clause_(pd,in);

        if(!t1)
        {
            return 0;
        }

        skip_space(in);

        if(**in == ']')
        {
            skip_one(in);
            void *c = pd->make_collection();
            pd->add_arg(c,t1);
            return pd->make_list(c);
        }

        if(**in == '|')
        {
            skip_one(in);

            void *t2 = parse_clause_(pd,in);

            if(!t2)
            {
                pd->destroy_any(t1);
                return 0;
            }

            skip_space(in);

            if(**in != ']')
            {
                pd->destroy_any(t1);
                pd->destroy_any(t2);
                return 0;
            }

            skip_one(in);
            return pd->make_split(t1,t2);
        }

        void *c = pd->make_collection();
        pd->add_arg(c,t1);

        for(;;)
        {
            skip_space(in);

            if(**in != ',')
            {
                pd->destroy_any(c);
                return 0;
            }

            skip_one(in);

            void *a = parse_clause_(pd,in);

            if(!a)
            {
                pd->destroy_any(c);
                return 0;
            }

            pd->add_arg(c,a);
            skip_space(in);

            if(**in == ']')
            {
                skip_one(in);
                return pd->make_list(c);
            }
        }
    }

#if defined(WIN32) || defined(__linux__)
	int digittoint(char a)
	{
		int ret = 0;
		if (a >= '0' && a<= '9')
		{
			ret = a - '0';
		}
		else if( a >= 'a' && a<= 'f' )
		{
			ret = a - 'a' + 10;
		}
		else if( a >= 'A' && a<= 'F' )
		{
			ret = a - 'A' + 10;
		}
		
		return ret;
	}
#endif
	
    void *parse_number_(pd_t *pd, const char **in)
    {
        bool neg = false;
        long n1 = 0;

        skip_space(in);

        if(**in == '-')
        {
            (*in)++;
            neg = true;
        }
        else if(**in == '+')
        {
            (*in)++;
        }

        if(**in != '.' && (**in <'0' && **in > '9'))
        {
            return 0;
        }

        while(**in >= '0' && **in<='9')
        {
            n1 = 10*n1+(**in)-'0';
            (*in)++;
        }

        if(**in != '.' && **in != 'e' && **in != 'E' && **in != 'f' && **in != 'F')
        {
            if(neg) n1=-n1;
            return pd->make_long(n1);
        }

        if(**in == 'f' || **in == 'F')
        {
            (*in)++;
            if(neg) n1=-n1;
            return pd->make_float(n1);
        }

        float f = float(n1);

        if(**in == '.')
        {
            (*in)++;

            float n2 = 0;
            unsigned digits = 0;

            while(**in >= '0' && **in<='9')
            {
                n2 = 10.0*n2+(**in)-'0';
                (*in)++;
                digits++;
            }

            if(digits) f += (float(n2)/pow((double)10.0,(double)digits));
        }

        if(neg) f=-f;

        if(**in != 'e' && **in != 'E')
        {
            if(**in == 'f' || **in == 'F') { (*in)++; }
            return pd->make_float(f);
        }

        (*in)++;

        long e = 0;
        neg = false;

        if(**in == '-')
        {
            (*in)++;
            neg = true;
        }
        else if(**in == '+')
        {
            (*in)++;
        }

        while(**in >= '0' && **in<='9')
        {
            e = 10*e+(**in)-'0';
            (*in)++;
        }

        if(**in == 'f' || **in == 'F') { (*in)++; }

        if(neg)
        {
            e = -e;
        }

        f = f*pow(10.0,e);
        return pd->make_float(f);
    }

    bool parse_string0_(const char **in, const char **wp, unsigned *wl, bool *esc)
    {
        skip_space(in);

        if(!**in)
        {
            return false;
        }

        char lead = **in;
        *esc = false;

        skip_one(in);

        const char *p = *in;

        while(**in && **in!=lead)
        {
            if(**in == '%') *esc = true;
            skip_one(in);
        }

        if(**in == lead)
        {
            skip_one(in);
            *wp = p;
            *wl = (*in)-p-1;
            return true;
        }

        return false;
    }

    void *parse_string_(pd_t *pd, const char **in)
    {
        const char *wp;
        unsigned wl;
        bool esc;

        if(!parse_string0_(in,&wp,&wl,&esc))
        {
            return 0;
        }

        if(!esc)
        {
            return pd->make_string(wp,wl);
        }

        char *es = (char *)malloc(wl);
        unsigned el=0;

        while(wl>0)
        {
            if(*wp!='%')
            {
                es[el++] = *wp++; wl--;
                continue;
            }

            wp++; wl--;

            if(*wp=='%')
            {
                es[el++] = '%'; wp++; wl--;
                continue;
            }

            if(wl<2 || !isxdigit(wp[0]) || !isxdigit(wp[1]))
            {
                free(es);
                return 0;
            }

            es[el++] = 16*digittoint(wp[0])+digittoint(wp[1]); wp+=2; wl-=2;
        }

        void *s = pd->make_string(es,el);
        free(es);
        return s;
    }

    bool parse_word_(const char **in, const char **w, unsigned *l)
    {
        skip_space(in);

        if(!is_word_start(**in))
        {
            return false;
        }

        *w = *in;
        (*in)++;

        while(is_word_body(**in))
        {
            (*in)++;
        }

        *l = *in-*w;
        return true;
    }

    void *parse_subst_(pd_t *pd, const char **in)
    {
        skip_space(in);

        if(**in != '~')
        {
            return 0;
        }

        skip_one(in);

        const char *wp;
        unsigned wl;

        if(**in != '(')
        {
            if(!parse_word_(in,&wp,&wl))
            {
                return 0;
            }

            return pd->make_subst(wp,wl,0,0);
        }

        skip_one(in);

        if(!parse_word_(in,&wp,&wl))
        {
            return 0;
        }

        skip_space(in);

        if(**in != ')')
        {
            return 0;
        }

        skip_one(in);
        skip_space(in);

        const char *w2p = *in;
        unsigned w2l;
        bool w2e;

        if(is_string_start(**in))
        {
            if(parse_string0_(in,&w2p,&w2l,&w2e))
            {
                return pd->make_subst(wp,wl,w2p,w2l);
            }

            return 0;
        }

        while(is_body(**in))
        {
            (*in)++;
        }

        w2l = (*in)-w2p;
        return pd->make_subst(wp,wl,w2l?w2p:0,w2l);
    }

    void *build_variable_(pd_t *pd, const char *wp, unsigned wl)
    {
        if(wl==4 && !strncmp(wp,"True",4)) return pd->make_bool(true);
        if(wl==5 && !strncmp(wp,"False",5)) return pd->make_bool(false);
        if(wl==4 && !strncmp(wp,"None",4)) return pd->make_none();
        return pd->make_variable(wp,wl);
    }

    void *build_atom_(pd_t *pd, const char *wp, unsigned wl)
    {
        if(wl==4 && !strncmp(wp,"True",4)) return pd->make_bool(true);
        if(wl==5 && !strncmp(wp,"False",5)) return pd->make_bool(false);
        if(wl==4 && !strncmp(wp,"None",4)) return pd->make_none();
        return pd->make_string(wp,wl);
    }

    void *parse_variable_(pd_t *pd, const char **in)
    {
        skip_space(in);

        if(!is_variable_start(**in))
        {
            return 0;
        }

        const char *p = *in;

        while(is_variable_body(**in))
        {
            (*in)++;
        }

        return build_variable_(pd,p,*in-p);
    }

    void *parse_clauselist_(pd_t *pd, const char **in)
    {
        void *t = pd->make_collection();

        for(;;)
        {
            const char *q = *in;

            void *a = parse_clause_(pd,in);

            if(!a)
            {
                *in = q;
                return t;
            }

            pd->add_arg(t,a);

            skip_space(in);

            if(**in != ',')
            {
                return t;
            }

            skip_one(in);
        }
    }

    void *parse_word_or_term_(pd_t *pd, const char **in, bool term)
    {
        skip_space(in);
        const char *wp;
        unsigned wl;

        if(!parse_word_(in,&wp,&wl))
        {
            return 0;
        }

        skip_space(in);

        if(**in != '(')
        {
            if(term)
            {
                return 0;
            }

            return build_atom_(pd,wp,wl);
        }

        skip_one(in);

        void *t = parse_clauselist_(pd,in);

        if(!t)
        {
            return 0;
        }

        skip_space(in);

        if(**in != ')')
        {
            pd->destroy_any(t);
            return 0;
        }

        skip_one(in);
        return pd->make_term(wp,wl,t);
    }

    void *parse_termlist_(pd_t *pd, const char **in)
    {
        void *t = pd->make_collection();

        for(;;)
        {
            const char *q = *in;

            void *a = parse_word_or_term_(pd,in,true);

            if(!a)
            {
                *in = q;
                return t;
            }

            pd->add_arg(t,a);

            skip_space(in);

            if(**in != ',')
            {
                return t;
            }

            skip_one(in);
        }
    }

    void *parse_clause_(pd_t *pd, const char **in)
    {
        skip_space(in);

        if(**in == '[') return parse_list_(pd,in);
        if(is_string_start(**in)) return parse_string_(pd,in);
        if(is_number_start(**in)) return parse_number_(pd,in);
        if(is_variable_start(**in)) return parse_variable_(pd,in);
        if(**in == '~') return parse_subst_(pd,in);

        return parse_word_or_term_(pd,in,false);
    }

    struct parse_state_delegate: pd_t
    {
        virtual void destroy_any(void *t)
        {
            delete (term_t *)t;
        }

        virtual void *make_collection()
        {
            return new term_t(0);
        }

        virtual void add_arg(void *t, void *a)
        {
            ((term_t *)t)->append_arg(*(term_t *)a);
        }

        virtual void *make_term(const char *p, unsigned pl, void *c_)
        {
            term_t *c = (term_t *)c_;
            term_t *t = new term_t(String::fromUTF8(p,pl),c->arity());

            for(unsigned i=0;i<c->arity();i++)
            {
                t->set_arg(i,c->arg(i));
            }

            delete c;
            return t;
        }

        virtual void *make_list(void *c)
        {
            return c;
        }

        virtual void *make_split(void *, void *)
        {
            return 0;
        }

        virtual void *make_long(long v)
        {
            return new term_t(String((int64)v));
        }

        virtual void *make_float(float v)
        {
            return new term_t(String(v));
        }

        virtual void *make_bool(bool v)
        {
            return new term_t(String(v?1:0));
        }

        virtual void *make_none()
        {
            return new term_t();
        }

        virtual void *make_string(const char *s, unsigned sl)
        {
            return new term_t(String::fromUTF8(s,sl));
        }

        virtual void *make_variable(const char *, unsigned)
        {
            return 0;
        }

        virtual void *make_subst(const char *, unsigned, const char *, unsigned)
        {
            return 0;
        }
    };

    void *parse_clause(pd_t *pd, const char *input)
    {
        const char *input_copy = input;
        void *t = parse_clause_(pd,&input_copy);
        if(!t) return 0;
        skip_space(&input_copy);
        if(!*input_copy) return t;
        pd->destroy_any(t);
        return 0;
    }

    void *parse_term(pd_t *pd, const char *input)
    {
        const char *input_copy = input;
        void *t = parse_word_or_term_(pd,&input_copy,true);
        if(!t) return 0;
        skip_space(&input_copy);
        if(!*input_copy) return t;
        pd->destroy_any(t);
        return 0;
    }

    void *parse_clauselist(pd_t *pd, const char *input)
    {
        const char *input_copy = input;
        void *t = parse_clauselist_(pd,&input_copy);
        if(!t)
        {
            return 0;
        }
        skip_space(&input_copy);
        if(!*input_copy) return t;
        pd->destroy_any(t);
        return 0;
    }

    void *parse_termlist(pd_t *pd, const char *input)
    {
        const char *input_copy = input;
        void *t = parse_termlist_(pd,&input_copy);
        if(!t)
        {
            return 0;
        }
        skip_space(&input_copy);
        if(!*input_copy) return t;
        pd->destroy_any(t);
        return 0;
    }
};

term_t::term_t()
{
}

term_t::~term_t()
{
}

const term_t::impl_t *term_t::impl() const
{
    return impl_.get();
}

String term_t::render() const
{
    if(!impl_.get())
    {
        return "";
    }

    String stream;
    impl_->render(stream);
    return stream;
}

term_t::term_t(const String &v)
{
    impl_ = ReferenceCountedObjectPtr<impl_t>(new term_atom_t(v));
}

term_t::term_t(unsigned len)
{
    if(len)
        impl_ = ReferenceCountedObjectPtr<impl_t>(new term_list_t(len));
    else
        impl_ = ReferenceCountedObjectPtr<impl_t>(new term_list_t);
}

term_t::term_t(const String &pred,unsigned len)
{
    if(len)
        impl_ = ReferenceCountedObjectPtr<impl_t>(new term_pred_t(pred,len));
    else
        impl_ = ReferenceCountedObjectPtr<impl_t>(new term_pred_t(pred));
}

term_t::term_t(const term_t &t): impl_(t.impl_)
{
}

term_t &term_t::operator=(const term_t &t)
{
    impl_=t.impl_;
    return *this;
}

unsigned term_t::type() const
{
    return impl_.get()?impl_->type_:PIW_TERM_NULL;

}

unsigned term_t::arity() const
{
    return impl_.get()?impl_->arity():0;
}

String term_t::pred() const
{
    return impl_.get()?impl_->pred():String::empty;
}

String term_t::value() const
{
    return impl_.get()?impl_->value():String();
}

term_t term_t::arg(unsigned n) const
{
    return impl_.get()?impl_->arg(n):term_t();
}

void term_t::set_arg(unsigned n, const term_t &v)
{
    if(impl_.get())
    {
        impl_->set_arg(n,v);
    }
}

unsigned term_t::count() const
{
    return impl_.get()?impl_->getReferenceCount():0;
}

void term_t::reset()
{
    impl_ = nullptr;
}

void term_t::append_arg(const term_t &v)
{

    if(impl_.get())
    {
        impl_->append_arg(v);
    }
}

term_t parse_state_term(const String &v)
{
    parse_state_delegate d;   
    term_t *t = (term_t *)parse_term(&d,v.toUTF8());
    if(!t) return term_t();
    term_t t2(*t);
    delete t;
    return t2;
}
